src/bm3dornl/utils.py

```diff
@@ -0,0 +1,269 @@
+
+    This function computes a list of indices for patches that are within a given row and column distance from
+    the reference patch specified by `ref_index`. It only considers patches that have not been compared previously
+    (i.e., patches that are ahead of the reference patch in the list, ensuring the upper triangle of the comparison matrix).
+
+    Parameters
+    ----------
+    signal_patches : np.ndarray
+        Array containing the positions of all signal patches. Each position is represented as (row_index, column_index).
+    ref_index : int
+        Index of the reference patch in `signal_patches` for which candidates are being sought.
+    cut_off_distance : tuple
+        A tuple (row_dist, col_dist) specifying the maximum allowed distances in the row and column directions.
+
+    Returns
+    -------
+    list
+        A list of integers representing the indices of the candidate patches in `signal_patches` that are within
+        the `cut_off_distance` from the reference patch and are not previously compared (ensuring upper triangle).
+    """
+    num_patches = signal_patches.shape[0]
+    ref_pos = signal_patches[ref_index]
+    candidate_patch_ids = []
+
+    for i in range(ref_index + 1, num_patches):  # Ensure only checking upper triangle
+        if (
+            np.abs(signal_patches[i, 0] - ref_pos[0]) <= cut_off_distance[0]
+            and np.abs(signal_patches[i, 1] - ref_pos[1]) <= cut_off_distance[1]
+        ):
+            candidate_patch_ids.append(i)
+
+    return candidate_patch_ids
+
+
+@jit(nopython=True)
+def is_within_threshold(
+    ref_patch: np.ndarray, cmp_patch: np.ndarray, intensity_diff_threshold: float
+) -> bool:
+    """
+    Determine if the Euclidean distance between two patches is less than a specified threshold.
+
+    This function computes the Euclidean distance between two patches and checks if it is less than the provided
+    intensity difference threshold. It is optimized with Numba's JIT in nopython mode to ensure high performance.
+
+    Parameters
+    ----------
+    ref_patch : np.ndarray
+        The reference patch as a flattened array of intensities.
+    cmp_patch : np.ndarray
+        The comparison patch as a flattened array of intensities.
+    intensity_diff_threshold : float
+        The threshold below which the Euclidean distance between the patches is considered sufficiently small
+        for the patches to be deemed similar.
+
+    Returns
+    -------
+    bool
+        True if the Euclidean distance between `ref_patch` and `cmp_patch` is less than `intensity_diff_threshold`;
+        otherwise, False.
+
+    Example:
+    --------
+    >>> ref_patch = np.array([1, 2, 3])
+    >>> cmp_patch = np.array([1, 2, 5])
+    >>> threshold = 2.5
+    >>> is_within_threshold(ref_patch, cmp_patch, threshold)
+    False
+    """
+    return np.linalg.norm(ref_patch - cmp_patch) <= intensity_diff_threshold
```

Comment:
In my head this returns an ndarray. Don't you need to add an `np.allTrue`?

Reply:
`np.linalg.norm` returns a single number, here is an example:

```bash
cloud in 🌐 devgpu in bm3dornl on  add_src_code via 🐍 v3.11.9 via 🅒 bm3dornl took 41s 
❯ ipython
Python 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.22.2 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: ref_patch = np.random.random(8*8).reshape(8, 8)

In [3]: cmp_patch = np.random.random(8*8).reshape(8, 8)

In [4]: intensity_diff_threshold = 0.5

In [5]: np.linalg.norm(ref_patch - cmp_patch) <= intensity_diff_threshold
Out[5]: False

In [6]: 
```